### PR TITLE
ARLAS Explo with IAM tests

### PR DESCRIPTION
**Pull request by sylvaingaudan**
Implement pytests for testing the APIs of ARLAS Exploration stack in IAM Mode

Mecanism:
- a command line (`scripts/pytest/init_cli_users.sh`) adds in `/tmp/arlas-cli-tests.yaml` different users: owner, user, anonymous, orphan, etc.
- a python script (scripts/pytest/conftest.py) declare **_fixtures_**. Fixtures are called before the execution of tests. Each test specify which fixture is necessary.
- a python script (scripts/pytest/test_iam.py) defines the tests. It uses `parametrize` with a set of values: this means that the test will be called several times, each iteration with one value of the set. At the end, there's more than 80 different tests.

Note: passwords are set to `secret` with a dedicated shell script
Note: the database of users and orgs is reseted before each tests by calling a `docker exec db psql` script

### CODE REVIEW

File: scripts/pytest/test_iam.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                         [OWNER_ORG_1,  # owner org 1 can not invite in org 2
+                          USER_ORG_1,   # user  org 1 can not invite in org 2
+                          USER_ORG_2,   # user  org 2 can not invite in org 2
+                          ORPHAN,       # orphan      can not invite in org 2
+                          ORPHAN2,      # orphan2     can not invite in org 2
+                          ORPHAN3,])    # orphan3     can not invite in org 2
+def test_not_add_user_to_org_as_not_owner(run_as, fixture_init, fixture_org1_owner_and_user, fixture_org2_owner, fixture_orphans):
+    """Test: as foreigner owner, foreigner user, user or orphan (with and without org-filter), I can not add a user in org"""
+    oid, __, __ = fixture_org2_owner
+    with pytest.raises(SystemExit):
```
Comment (MohamedHamouGisaia) [inline scripts/pytest/test_iam.py:173]: shouldn't be USER_ORG_2 ?

File: scripts/pytest/conftest.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                    "track.timestamps.center": "date-epoch_second",
+                    "track.timestamps.start": "date-epoch_second",
+                    "track.timestamps.end": "date-epoch_second"},
+                no_fulltext=["cargo_type"])
+            Service.create_index(
+                USER_ADMIN,
+                index=index,
+                mapping=mapping)
+    os.system("docker exec db psql arlas -c \"" + SQL_INIT.replace(os.linesep, " ") + "\" > /dev/null")
+    collections = Service.list_collections(USER_ADMIN)[1:]
```
Comment (QuCMGisaia) [inline scripts/pytest/conftest.py:108]: What do those lines do ? It seems like they remove collection in the init

File: scripts/pytest/conftest.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    email = create_user(USER_ADMIN, OWNER_ORG_1)
+    Service.add_user_in_organisation(USER_ADMIN, oid, OWNER_ORG_1, list(groups.values()))
+    return oid, email, groups
+
+
+@pytest.fixture(scope="function")
+def fixture_org1_owner_and_user(fixture_org1_owner):
+    oid, email1, groups = fixture_org1_owner
+    email2 = create_user(OWNER_ORG_1, USER_ORG_1)
+    print(Service.add_user_in_organisation(OWNER_ORG_1, oid, USER_ORG_1, [groups.get("role/arlas/user"), groups.get("group/config.json/" + ORG_1)]))
```
Comment (QuCMGisaia) [inline scripts/pytest/conftest.py]: trailing print

File: scripts/pytest/conftest.py (hunk 2 above)
Comment (QuCMGisaia) [reply scripts/pytest/conftest.py:108]: It would make more sense to have it in the beginning of the collections method

File: scripts/pytest/test_iam.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+@pytest.mark.parametrize("run_as",
+                         [OWNER_ORG_2,  # owner org 2 can not see private collection org 1
+                          USER_ORG_2,   # user  org 2 can not see private collection org 1
+                          ANONYMOUS,    # anonymous   can not see private collection org 1
+                          ORPHAN,       # orphan      can not see private collection org 1
+                          ORPHAN2,      # orphan2     can not see private collection org 1
+                          ORPHAN3])     # orphan3     can not see private collection org 1
+def test_foreingers_does_not_see_private_collection(run_as, fixture_init, fixture_org1_owner_and_user, fixture_org2_owner_and_user, fixture_orphans, fixture_collections):
+    """Test: as foreing user, I can not see the collections of other orgs"""
```
Comment (QuCMGisaia) [inline scripts/pytest/test_iam.py]: foreign

File: scripts/pytest/test_iam.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+# Test collection describe visibility
+
+@pytest.mark.parametrize("run_as",
+                         [OWNER_ORG_2,  # owner org 2 can not see private collection org 1
+                          USER_ORG_2,   # user  org 2 can not see private collection org 1
+                          ANONYMOUS,    # anonymous   can not see private collection org 1
+                          ORPHAN,       # orphan      can not see private collection org 1
+                          ORPHAN2,      # orphan2     can not see private collection org 1
+                          ORPHAN3])     # orphan3     can not see private collection org 1
+def test_foreingers_does_not_see_private_collection_description(run_as, fixture_init, fixture_org1_owner_and_user, fixture_org2_owner_and_user, fixture_orphans, fixture_collections):
```
Comment (QuCMGisaia) [inline scripts/pytest/test_iam.py]: foreigners

Comment (QuCMGisaia) [reply scripts/pytest/test_iam.py]: does -> do

File: scripts/pytest/test_iam.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+# Test collection listing visibility
+
+@pytest.mark.parametrize("run_as",
+                         [OWNER_ORG_2,  # owner org 2 can not see private collection org 1
+                          USER_ORG_2,   # user  org 2 can not see private collection org 1
+                          ANONYMOUS,    # anonymous   can not see private collection org 1
+                          ORPHAN,       # orphan      can not see private collection org 1
+                          ORPHAN2,      # orphan2     can not see private collection org 1
+                          ORPHAN3])     # orphan3     can not see private collection org 1
+def test_foreingers_does_not_see_private_collection(run_as, fixture_init, fixture_org1_owner_and_user, fixture_org2_owner_and_user, fixture_orphans, fixture_collections):
```
Comment (QuCMGisaia) [inline scripts/pytest/test_iam.py]: foreigners_do

File: scripts/pytest/test_iam.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+# Test collection sample visibility
+
+@pytest.mark.parametrize("run_as",
+                         [OWNER_ORG_2,  # owner org 2 can not see private collection org 1
+                          USER_ORG_2,   # user  org 2 can not see private collection org 1
+                          ANONYMOUS,    # anonymous   can not see private collection org 1
+                          ORPHAN,       # orphan      can not see private collection org 1
+                          ORPHAN2,      # orphan2     can not see private collection org 1
+                          ORPHAN3])     # orphan3     can not see private collection org 1
+def test_foreingers_does_not_see_private_collection_sample(run_as, fixture_init, fixture_org1_owner_and_user, fixture_org2_owner_and_user, fixture_orphans, fixture_collections):
```
Comment (QuCMGisaia) [inline scripts/pytest/test_iam.py]: foreigners_do

**Review by QuCMGisaia — APPROVED** [review]

File: scripts/pytest/conftest.py (hunk 2 above)
Comment (sylvaingaudan) [reply scripts/pytest/conftest.py:108]: The fixture_init **initialize** the arlas stack: it empties everything if it wasn't clean. This is why it:
- empties the sql tables
- empties the collections

Comment (sylvaingaudan) [reply scripts/pytest/conftest.py:108]: _"It would make more sense to have it in the beginning of the collections method"_

=> I do not necessarily agree: having a clean stack is the role of that fixture.

File: scripts/pytest/conftest.py (hunk 3 above)
Comment (sylvaingaudan) [reply scripts/pytest/conftest.py]: yes, thank you!

File: scripts/pytest/test_iam.py (hunk 1 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py:173]: In fact, doing for ORPHAN is even better, more combinaision tested. Thanks

File: scripts/pytest/test_iam.py (hunk 6 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: thanks

File: scripts/pytest/test_iam.py (hunk 4 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: thanks

File: scripts/pytest/test_iam.py (hunk 7 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: thanks

File: scripts/pytest/test_iam.py (hunk 5 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: ok

File: README.md [hunk 8]
```diff
@@ -19,3 +19,14 @@ To release, run:
...
 To update the version of the dependencies, such as ARLAS containers, edit `conf/versions.env`.
+
+## Tests
+
+Launch tests for IAM:
+
+```shell
+./start.sh iam
+pip3.10 install arlas_cli
+pip.3.10 install pytest==8.3.4
```
Comment (WilliGautier) [inline README.md]: pip3.10

File: README.md [hunk 9]
```diff
@@ -19,3 +19,14 @@ To release, run:
...
+pytest -s scripts/pytest/test_iam.py
```
Comment (WilliGautier) [inline README.md:31]: Missing 
`./scripts/pytest/init_cli_users.sh`

File: scripts/pytest/test_iam.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                                         ]
+
+CAN_SEE_COLLECTIONS = [[OWNER_ORG_1, COLLECTION1_ORG_1_PRIVATE],               # owner org 1 can see private collection org 1
+                       [USER_ORG_1, COLLECTION1_ORG_1_PRIVATE],                # user  org 1 can see private collection org 1
+                       [OWNER_ORG_2, COLLECTION1_ORG_1_PUBLIC],                # owner org 2 can see public  collection org 1
+                       [USER_ORG_2, COLLECTION1_ORG_1_PUBLIC],                 # user  org 1 can see public  collection org 1
+                       [USER_ORG_2_NO_ORG_FILTER, COLLECTION1_ORG_1_PUBLIC],   # user  org 2 can see private collection org 1
+                       [ANONYMOUS, COLLECTION1_ORG_1_PUBLIC],                  # anonymous   can see public  collection org 1
+                       [ORPHAN, COLLECTION1_ORG_1_PUBLIC],                     # orphan      can see public  collection org 1
+                       [ORPHAN_ORG_FILTER_ORG1, COLLECTION1_ORG_1_PUBLIC],     # orphan      can see public  collection org 1
```
Comment (WilliGautier) [inline scripts/pytest/test_iam.py:55]: The numerotation of ORPHANs helps to identify them, could be added to the comments (as in the `CAN_NOT_SEE_FOREIGN_PRIVE_COLLECTIONS` before) and even in variable name:
`ORPHAN_ORG_FILTER_ORG1` -> `ORPHAN2_ORG_FILTER_ORG1`

File: scripts/pytest/test_iam.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+def test_orphan_not_create_org(fixture_init, fixture_orphans):
+    orphan, __, __, __, __, __ = fixture_orphans
+    with pytest.raises(SystemExit):
+        assert Service.create_organisation(orphan, ORG_1)
+
+
+# Test add users
+def test_add_user_to_org_as_admin(fixture_init, fixture_org1_owner):
+    """as admin, I can add a user in org"""
+    oid, email, groups = fixture_org1_owner
```
Comment (WilliGautier) [inline scripts/pytest/test_iam.py]: The user should be added no?

File: README.md (hunk 8 above)
Comment (sylvaingaudan) [reply README.md]: indeed. Thanks

File: scripts/pytest/test_iam.py (hunk 11 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: It's done in the fixture, I check here that the fixture did its job. Ok with you?

File: scripts/pytest/test_iam.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+                             ]
+
+
+# Test organisation creation
+def test_admin_create_org(fixture_init):
+    assert Service.create_organisation(USER_ADMIN, ORG_1).get("name") == ORG_1
+
+
+def test_orphan_not_create_org(fixture_init, fixture_orphans):
+    orphan, __, __, __, __, __ = fixture_orphans
```
Comment (mbarbet) [inline scripts/pytest/test_iam.py]: fixture_orphans return s5 elements
here we have 6
you need to remove one __

Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: indeed. Thank you.

File: scripts/pytest/test_iam.py (hunk 11 above)
Comment (WilliGautier) [reply scripts/pytest/test_iam.py]: Ah ok

File: scripts/pytest/test_iam.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    with pytest.raises(SystemExit):
+        assert Service.create_organisation(orphan, ORG_1)
```
Comment (WilliGautier) [inline scripts/pytest/test_iam.py]: If it has to catch the `SystemExit`, the assert should be removed?

File: README.md (hunk 9 above)
Comment (sylvaingaudan) [reply README.md:31]: yes thank you. Actually, this is going to move in the python.

File: scripts/pytest/test_iam.py [hunk 14]
```diff
@@ -54,124 +75,431 @@
...
     assert Service.create_organisation(USER_ADMIN, ORG_1).get("name") == ORG_1
 
 
-def test_orphan_not_create_org(fixture_init, fixture_orphans):
-    orphan, __, __, __, __, __ = fixture_orphans
+@pytest.mark.parametrize("run_as", NON_ADMIN)
+def test_non_admin_not_create_org(run_as, cli_confs, fixture_init, fixture_org2_owner_and_users, fixture_orphans):
+    """Test: as not admin, I can not create an org"""
+    if run_as == ANONYMOUS:  # arlas_cli does not handle calls to iam without identity
+        requests.post("https://localhost/arlas_iam_server/organisations/" + ORG_1, headers={"accept": "application/json;charset=utf-8"}, verify=False).status_code > 299
```
Comment (QuCMGisaia) [inline scripts/pytest/test_iam.py]: Why not use `anonymous_iam_call` instead of the `requests.post` ?

**Review by QuCMGisaia — APPROVED** [review]

File: scripts/pytest/test_iam.py (hunk 13 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: ok

File: scripts/pytest/test_iam.py (hunk 14 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: indeed. Thnaks

File: scripts/pytest/conftest.py [hunk 15]
```diff
@@ -163,6 +170,8 @@ def fixture_collections():
     create_collection(OWNER_ORG_1, ORG_1, COLLECTION1_ORG_1_PRIVATE, INDEX_ORG1, False)
     create_collection(OWNER_ORG_1, ORG_1, COLLECTION1_ORG_1_PUBLIC, INDEX_ORG1, True)
     create_collection(OWNER_ORG_2, ORG_2, COLLECTION1_ORG_2_PRIVATE, INDEX_ORG2, False)
+    time.sleep(1)
     Service.describe_collection(OWNER_ORG_1, COLLECTION1_ORG_1_PRIVATE)
     Service.describe_collection(OWNER_ORG_1, COLLECTION1_ORG_1_PUBLIC)
     Service.describe_collection(OWNER_ORG_2, COLLECTION1_ORG_2_PRIVATE)
+    time.sleep(1)
```
Comment (QuCMGisaia) [inline scripts/pytest/conftest.py:179]: Is this sleep really needed ? Why would ARLAS need this one second before the setup of each test ? Removing it would save about 5min in the total CI time

File: scripts/pytest/test_iam.py [hunk 16]
```diff
@@ -0,0 +1,519 @@
...
+    if run_as == ANONYMOUS:  # arlas_cli does not handle calls to iam without identity
+        anonymous_iam_call("organisations/" + ORG_1, post={}).status_code > 299
+    else:
+        with pytest.raises(SystemExit):
+            Service.create_organisation(run_as, ORG_1)
+
+
+def test_anonymous_not_create_org_on_user_domain(fixture_cli_confs, fixture_init, fixture_orphans):
+    """Test: as anonymous, I can not create an org on my domain"""
+    requests.post("https://localhost/arlas_iam_server/organisations", headers={"accept": "application/json;charset=utf-8"}, verify=False).status_code > 299  # NOSONAR
```
Comment (QuCMGisaia) [inline scripts/pytest/test_iam.py]: Use anonymous_iam_call

**Review by QuCMGisaia — APPROVED** [review]

File: scripts/pytest/conftest.py (hunk 15 above)
Comment (sylvaingaudan) [reply scripts/pytest/conftest.py:179]: because on the CI, the collection is created but not instantly visible from the users (arlas server cache?). `Service.describe_collection` combined with the `time.sleep(1)` makes the collection visible. Would you have another mean?

Comment (QuCMGisaia) [reply scripts/pytest/conftest.py:179]: Hmm not really, I find it surprising that the collection is not directly visible. Is there any acknowledgment sent by the server once the request is fully processed ? Shouldn't we wait for the `describe_collection` method to end if it is async ?

Comment (sylvaingaudan) [reply scripts/pytest/conftest.py:179]: I do already wait for the end of `describe_collection`. But as another user (the one in the test), I get an empty list of   collection, then a second after I see it. I have difficulties to explain it (maybe the "eventually consistent" property of elasticsearch?).

File: scripts/pytest/test_iam.py (hunk 16 above)
Comment (sylvaingaudan) [reply scripts/pytest/test_iam.py]: indeed, thanks

File: scripts/pytest/conftest.py (hunk 15 above)
Comment (sylvaingaudan) [reply scripts/pytest/conftest.py:179]: the problem is arlas server, not the test.
